Fix persistent worker exits before pin_memory thread

ejguan opened the pull request:
Fixes #1551

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#71579 Fix persistent worker exits before pin_memory thread**

As the comment in the code, register a function to terminate persistent workers. 
By adding a reference of these workers in `atexit`, it would prevent Python interpreter kills these persistent worker processes before `pin_memorh_thread` exits.
And, if users explicitly kills DataLoader iterator, such function in `atexit` would be a no-op.

Differential Revision: [D33896537](https://our.internmc.facebook.com/intern/diff/D33896537)

ejguan committed:
Fix persistent worker exists before pin_memory thread

[ghstack-poisoned]

ejguan committed:
Update on "Fix persistent worker exists before pin_memory thread"

Fixes #1551


As the comment in the code, adding a function to manually delete the iterator with persistent worker. It would invoke `__del__` inside Iterator object and make sure pin_memory_thread exits before worker process.

I choose using `atexit` rather than adding `__del__` to DataLoader because it's not safe as the destructor function may not be invoked when Python interpreter exits.

[ghstack-poisoned]

ejguan committed:
Update on "Fix persistent worker exits before pin_memory thread"

Fixes #1551


As the comment in the code, adding a function to manually delete the iterator with persistent worker. It would invoke `__del__` inside Iterator object and make sure pin_memory_thread exits before worker process.

I choose using `atexit` rather than adding `__del__` to DataLoader because it's not safe as the destructor function may not be invoked when Python interpreter exits.

[ghstack-poisoned]

ejguan commented on torch/utils/data/dataloader.py:
```diff
@@ -946,6 +946,18 @@ def __init__(self, loader):
         else:
             self._data_queue = self._worker_result_queue
 
+        # In some rare cases, persistent workers (daemonic processes)
+        # would be terminated before `__del__` of iterator is invoked
+        # when main process exits
+        # It would cause failure when pin_memory_thread tries to read
+        # corrupted data from worker_result_queue
+        # atexit is used to prevent persistent workers exiting before
+        # pin_memory_thread
```
Here is the comment.

ejguan commented on test/test_dataloader.py:
```diff
@@ -2302,6 +2302,41 @@ def test_dataset_not_reset(self):
                 # and can cache values safely
                 dataset.start = i
 
+    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
+    def test_early_exit(self):
+        import subprocess
+        proc = subprocess.run([sys.executable, '-c', """\
+import torch
+from torch.utils.data import DataLoader, IterableDataset
+
+class RandomDataset(IterableDataset):
+    def __init__(self, len, size):
+        super(RandomDataset).__init__()
+        self.len = len
+        self.size = size
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.len <= 0:
+            raise StopIteration
+        self.len -= 1
+        return torch.randn(self.size)
+
+if __name__ == '__main__':
+    dl = DataLoader(
+        RandomDataset(64, (28, 28)),
+        batch_size=16,
+        num_workers=2,
+        pin_memory=True,
+        persistent_workers=True,
+    )
+
+    for _ in dl:
+        break
+"""], stderr=subprocess.PIPE)
+        self.assertTrue(proc.stderr == b'')
```
This test won't pass before this PR.

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33694867).

ejguan commented on torch/utils/data/dataloader.py:
```diff
@@ -946,6 +946,18 @@ def __init__(self, loader):
+        # corrupted data from worker_result_queue
+        # atexit is used to prevent persistent workers exiting before
+        # pin_memory_thread
+        if self._persistent_workers and self._pin_memory:
```
In order to potentially fix part of #71187 for the case of persistent workers, we may want to remove this condition on `self._pin_memory`

malfet commented:
Reverting as it broke Windows test sanity(from https://github.com/pytorch/pytorch/runs/4903195822?check_suite_focus=true):
```
Traceback (most recent call last):
  File "test_dataloader.py", line 2339, in test_early_exit
    self.assertTrue(proc.stderr == b'')
AssertionError: False is not true```

ejguan committed:
Update on "Fix persistent worker exits before pin_memory thread"

Fixes #1551


As the comment in the code, register a function to terminate persistent workers. 
By adding a reference of these workers in `atexit`, it would prevent Python interpreter kills these persistent worker processes before `pin_memorh_thread` exits.
And, if users explicitly kills DataLoader iterator, such function in `atexit` would be a no-op.


[ghstack-poisoned]

ejguan committed:
Update on "Fix persistent worker exits before pin_memory thread"

Fixes #1551


As the comment in the code, register a function to terminate persistent workers. 
By adding a reference of these workers in `atexit`, it would prevent Python interpreter kills these persistent worker processes before `pin_memorh_thread` exits.
And, if users explicitly kills DataLoader iterator, such function in `atexit` would be a no-op.


[ghstack-poisoned]

ejguan committed:
Update on "Fix persistent worker exits before pin_memory thread"

Fixes #1551


As the comment in the code, register a function to terminate persistent workers. 
By adding a reference of these workers in `atexit`, it would prevent Python interpreter kills these persistent worker processes before `pin_memorh_thread` exits.
And, if users explicitly kills DataLoader iterator, such function in `atexit` would be a no-op.


[ghstack-poisoned]

ejguan committed:
Update on "Fix persistent worker exits before pin_memory thread"

Fixes #1551


As the comment in the code, register a function to terminate persistent workers. 
By adding a reference of these workers in `atexit`, it would prevent Python interpreter kills these persistent worker processes before `pin_memorh_thread` exits.
And, if users explicitly kills DataLoader iterator, such function in `atexit` would be a no-op.


[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33896537).

ejguan committed:
Update on "Fix persistent worker exits before pin_memory thread"

Fixes #1551


As the comment in the code, register a function to terminate persistent workers. 
By adding a reference of these workers in `atexit`, it would prevent Python interpreter kills these persistent worker processes before `pin_memorh_thread` exits.
And, if users explicitly kills DataLoader iterator, such function in `atexit` would be a no-op.

Differential Revision: [D33896537](https://our.internmc.facebook.com/intern/diff/D33896537)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33896537).

anjali411 commented:
@ejguan can you also put a topic label to it? 😊
